Reset memoize to support streaming multiple objects.

jeroenbakker-atmind opened the pull request:
By default (value_)from_reader closes the input stream. It is possible
to deserialize multiple pickle objects from a single stream by
implementing a custom reader.

This patch will reset the memoize at the beginning when a new object is
deserialized. Any refs would then point to the correct object inside the memoize.

We checked several approaches but this approach seems to fit best (support most
protocols) however this might not be the obvious place.

jbakker-atmind committed:
Added support for streaming multiple pickles from a single stream.

When used to deserialize multiple pickle dumps from a single
stream the internal state needs to be reset. When not done
the instance indexes (memo) will be incorrect.

jbakker-atmind committed:
Added documentation how to reuse a pickle stream.

m-dewanchand committed:
Merge branch 'master' into feature-reset-deserializer-state

m-dewanchand committed:
Added options io boolean when creating new deserializer.

birkenfeld commented on src/de.rs:
```diff
@@ -143,6 +143,10 @@ impl<R: Read> Deserializer<R> {
     fn parse_value(&mut self) -> Result<Value> {
+        // Clear memo, to allow reading multiple pickle dump calls to
+        // a single stream.
+        self.memo.clear();
```
Are you sure this is correct here? `parse_value` is not only the entry point, it is also called at other points during deserialization.

jeroenbakker-atmind replied:
We tried several solutions
* Use a public method to reset the memo.
* Reset the memo when a `PROTO` byte was read. This would not work for older protocols as they don't send it.
* At `STOP` doesn't work as the memo is used afterwards.
* This location, but seems like it could be called with `variant_seed` so isn't the correct location.

So perhaps the valid solution is to use a public method? Do you have any other ideas?

birkenfeld commented on src/lib.rs:
```diff
@@ -64,6 +64,49 @@
 //! handle).  These functions, called `value_from_*` and `value_to_*`, will
 //! correctly (un)pickle these types.
 //!
+//! By default `(value_)from_reader` closes the input stream. It is possible
```
I'd prefer this new chunk to go to the `from_reader` docs, it is a bit too specialized for the crate intro doc.

jbakker-atmind committed:
Moved documentation from lib header to de.

Solves one issue of the code review.
The docs is put in the header of de.rs as it fits better than
`(value_)from_reader`.

jbakker-atmind committed:
Use public method to reset the internal state.

jbakker-atmind committed:
Reverted unneeded change.

jeroenbakker-atmind commented on src/de.rs:
```diff
@@ -9,6 +9,7 @@
 //! Note: Serde's interface doesn't support all of Python's primitive types.  In
 //! order to deserialize a pickle stream to `value::Value`, use the
 //! `value_from_*` functions exported here, not the generic `from_*` functions.
+//!
```
remove this unneeded change.

birkenfeld commented:
Thanks, I have no objections to this version. The doctests for 1.41.1 seem to fail because a `Cursor` of an array isn't `Read` - could you make the input a slice in the example?

jeroenbakker-atmind commented:
Yes, we will check on 1.41.1

jbakker-atmind committed:
Fix doctest not working in rust 1.41.1.

birkenfeld commented:
Thanks!
